include/aikido/planner/CompositePlanner.hpp

```diff
@@ -0,0 +1,62 @@
+#ifndef AIKIDO_PLANNER_COMPOSITEPLANNER_HPP_
+#define AIKIDO_PLANNER_COMPOSITEPLANNER_HPP_
+
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+#include "aikido/planner/Planner.hpp"
+
+namespace aikido {
+namespace planner {
+
+/// CompositePlanner is a base class for concrete planner classes that contain
+/// multiple planners.
+class CompositePlanner : public Planner
+{
+public:
+  /// Constructor.
+  CompositePlanner();
+
+  /// Constructs given list of planners.
+  explicit CompositePlanner(const std::vector<PlannerPtr>& planners);
+
+  /// Adds planner.
+  void addPlanner(PlannerPtr planner);
+
+  /// Returns true if this CompositePlanner contains \c planner.
+  bool hasPlanner(const Planner* planner);
+
+  /// Returns all the planners in this CompositePlanner.
+  const std::vector<PlannerPtr>& getPlanners() const;
+
+  /// Returns planner given \c index.
+  ///
+  /// \throw If index is equal to or greater than the number of planners in this
+  /// CompositePlanner.
+  PlannerPtr getPlanner(std::size_t index);
```

Comment:
Your inference is correct. Let me add `getNumPlanners()`.

One additional rationale behind this is for const-correctness (again :laughing:). Since this class already has `getPlanners()`, it can be seemed `getPlanner()` is not necessary. But, `getPlanners()` should be non-const because it returns `std::vector<PlannerPtr>`, which is a list of non-const instances. We can't call this function from const instance of `CompositePlanner` unless we additionally maintain `std::vector<ConstPlannerPtr>`. This is a considerable approach, but for now, let's have const/non-const `getPlanner()` for simplicity.

Reply:
Sorry, I think I'm still missing something. Are you saying that `getPlanner` and `getNumPlanners` are necessary in order for a meta-planner to work? (And we aren't able to just iterate over the vector of planners because of this const-correctness issue? Why can't we just always have (one) vector of `ConstPlanner`s? Maybe a related question: when would we want to change a `Planner`?)

Reply:
There _can_ be instances when the planner is changing its parameters as it plans. If these params are modified as member variables, we might need `<non-const>Planner`. :thinking:

Reply:
@aditya-vk I also considered that case at first, but now I think the instances being changed as it plans should be mutable. This is because I think planner logically shouldn't need to change its internal states but just the returning trajectory and the result.

So I'm thinking of having const planners, and then having only `getPlanners()` should be okay.

Reply:
Since we decided not to make `plan()` const, this function shouldn't be removed for these reasons:
* `getPlanner(index) const` for const iteration
* `getPlanner(index)` to provide consistent API with `getPlanner(index) const`.

But I'm open to removing `getPlanner(index)` because we could still do non-const iteration using `getPlanners()`.

Reply:
If I remember our discussion correctly, I think we concluded that external users might not need to be able to iterate over the individual planners at all? I'm in favor of removing `getPlanner(index)` and `getNumPlanners`, and potentially adding them back if we find they're useful or necessary.